source/base/MessageHandler.cpp

```diff
@@ -61,52 +61,124 @@ void MessageHandler::handleMessageReceived(Swift::Message::ref message)
+                    using namespace boost::posix_time;
+                    static ptime epoch(boost::gregorian::date(1970, 1, 1));
+                    time_duration diff(forwarded->getDelay()->getStamp() - epoch);
+                    timestamp = diff.ticks() / diff.ticks_per_second();
+                }
+            }
         }
     }
 
-    std::string fromJid = message->getFrom().toBare().toString();
-
-    // XEP 280
-    Swift::JID toJID;
-    bool sentCarbon = false;
-    // If this is a carbon message, we need to retrieve the actual content
-    if (settings_->getJid().compare(QString::fromStdString(fromJid)) == 0) {
+    // XEP 280 Carbon Messages
+    if ((isMamMsg == false) && (settings_->getJid().compare(partyJid) == 0)) {
         Swift::CarbonsReceived::ref carbonsReceived;
         Swift::CarbonsSent::ref carbonsSent;
-        Swift::Forwarded::ref forwarded;
-        Swift::Message::ref forwardedMessage;
         if ((carbonsReceived = message->getPayload<Swift::CarbonsReceived>()) &&
                 (forwarded = carbonsReceived->getForwarded()) &&
                 (forwardedMessage = std::dynamic_pointer_cast<Swift::Message>(forwarded->getStanza()))) {
             // It is the carbon of a message that we received
             message = forwardedMessage;
-            fromJid = message->getFrom().toBare().toString();
+            partyJid = QString::fromStdString(message->getFrom().toBare().toString());
+            qDebug() << "Carbon received" << endl;
         }
         else if ((carbonsSent = message->getPayload<Swift::CarbonsSent>()) &&
                  (forwarded = carbonsSent->getForwarded()) &&
                  (forwardedMessage = std::dynamic_pointer_cast<Swift::Message>(forwarded->getStanza()))) {
             // It is the carbon of a message that we sent
             message = forwardedMessage;
-            toJID = forwardedMessage->getTo();
-            sentCarbon = true;
+            partyJid = QString::fromStdString(forwardedMessage->getTo().toString());
+            qDebug() << "Carbon sent" << endl;
+        }
+        else
+        {
+            qDebug() << "Error Carbon" << endl;
         }
     }
 
-    boost::optional<std::string> fromBody = message->getBody();
+    QString resource = QString::fromStdString(message->getFrom().getResource());
+
+    // XEP 45 MUC messages
+    bool isGroupMessage = false;
+    if(message->getType() == Swift::Message::Groupchat)
+    {
+        qDebug() << "Group message" << endl;
+        isGroupMessage = true;
+        auto mucUser = message->getPayload<Swift::MUCUserPayload>();
+        if(mucUser)
+        {
+            auto items = mucUser->getItems();
+            if(items.size() > 0 && items[0].realJID)
+            {
+                auto msgOwnerJid = items[0].realJID;
+                direction = msgOwnerJid->compare(clientBareJid, Swift::JID::WithoutResource) == 0 ? 0 : 1;
+                partyJid = QString::fromStdString(msgOwnerJid->toBare().toString());
```

Comment:
```suggestion
                partyJid = QString::fromStdString(message->getFrom().toBare().toString());
```

Reply:
This fixes the issue for me, that the mam group messages are not in the group room, but associated to the sender of the message.